src/main/java/com/icthh/xm/ms/entity/service/processor/FormSpecProcessor.java

```diff
@@ -0,0 +1,165 @@
+                updatedSpecForm = resolveReferences(specifications, updatedSpecForm);
+            }
+            log.warn("Maximum iteration limit reached: {}. Could be recursion and form has not been processed", POSSIBLE_RECURSIVE_LIMIT);
+        }
+    }
+
+    private Map<String, String> collectTenantSpecifications(Set<String> existingReferences, String tenant) {
+        Map<String, String> specificationsByRelativePath = new LinkedHashMap<>();
+
+        for (String formPath : existingReferences) {
+            String formKey = matcher.extractUriTemplateVariables(KEY_FORM_TEMPLATE, formPath).get("formKey");
+
+            if (StringUtils.isNotBlank(formKey)) {
+                ofNullable(formsByTenant.get(tenant))
+                    .map(x -> x.get(formKey))
+                    .map(formSpec -> getFormSpecificationByFile(tenant, formSpec))
+                    .filter(StringUtils::isNotBlank)
+                    .ifPresentOrElse(
+                        specification -> specificationsByRelativePath.put(formKey, specification),
+                        () -> log.warn("The form specification for key:{} and tenant:{} was not found.", formKey, tenant));
+            }
+        }
+        return specificationsByRelativePath;
+    }
+
+    private String resolveReferences(Map<String, String> specifications, String typeSpecForm) throws JsonProcessingException {
+        ObjectMapper objectMapper = new ObjectMapper();
+        JsonNode defaults = objectMapper.readValue(typeSpecForm, JsonNode.class);
+        JsonNode jsonNode = replaceReferences(defaults, specifications);
+
+        return objectMapper.writeValueAsString(jsonNode);
+    }
+
+    private String getFormSpecificationByFile(String tenant, FormSpec formSpec) {
+        return ofNullable(formSpec.getValue())
+            .orElseGet(() -> jsonListenerService.getSpecificationByTenantRelativePath(tenant, formSpec.getRef()));
+    }
+
+    private JsonNode replaceReferences(JsonNode node, Map<String, String> specifications) {
+        if (node.isObject()) {
+            ObjectNode objectNode = (ObjectNode) node;
+            Iterator<Map.Entry<String, JsonNode>> it = convertToFailSafeIterator(objectNode.fields());
+
+            while (it.hasNext()) {
+                Map.Entry<String, JsonNode> entry = it.next();
+                String jsonKey = entry.getKey();
+                JsonNode value = entry.getValue();
+                if (jsonKey.equals(REF)) {
+                    String textValue = entry.getValue().asText();
+                    objectNode.remove(REF);
+
+                    specifications.keySet()
+                        .stream()
+                        .filter(textValue::contains)
+                        .findAny()
+                        .ifPresent(key -> objectNode.setAll(
+                            convertSpecificationToObjectNode(specifications.get(key))));
```

Comment:
better to add addition map function that transform
```key >>> specifications.get(key) >>> convertSpecificationToObjectNode(<<<>>>) >>> objectNode::setAll```

Reply:
Seems pretty good. Thanks.